sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateTableExec.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalog.v2.{Identifier, TableCatalog}
+import org.apache.spark.sql.catalog.v2.expressions.Transform
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.execution.LeafExecNode
+import org.apache.spark.sql.types.StructType
+
+case class CreateTableExec(
+    catalog: TableCatalog,
+    identifier: Identifier,
+    tableSchema: StructType,
+    partitioning: Seq[Transform],
+    tableProperties: Map[String, String],
+    ignoreIfExists: Boolean) extends LeafExecNode {
+  import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    if (!catalog.tableExists(identifier)) {
+      try {
+        catalog.createTable(identifier, tableSchema, partitioning.toArray, tableProperties.asJava)
```

Comment:
I checked presto's SPI, it asks the connector to implement a [`createTable`](https://github.com/prestodb/presto/blob/3060c65a1812c6c8b0c2ab725b0184dbad67f0ed/presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java#L182) method with the `ignoreExisting` parameter.

When implementing Hive/JDBC with data source v2, I think it's better to directly pass the `ignoreIfExists` flag, as these data sources support this flag natively.

I don't know the exact reason why presto designed its SPI in this way, maybe it's because the data source can have a chance to optimize for the `ignoreIfExists` flag. I think it's better to follow the design of presto here.

Reply:
BTW I think it's a separated issue from adding CREATE TABLE support. I'm fine as long as we add a TODO here.

Reply:
@cloud-fan, adding an additional argument to the `createTable` method is a poor choice because it forces Spark to depend on sources to implement consistent behavior. Consistency and reliability is a problem in Spark that we are trying to address by making Spark handle these cases.

That's why not adding a flag to `createTable` is the right choice. It keeps the API simpler for implementers and guarantees consistent behavior.

Reply:
The question is if the downstream source has different behavior from what Spark wants to enforce if the `ignoreIfExists` flag is passed to the source vs. Spark deciding how to handle it. So I can imagine there being a discrepancy if the user gets different behavior from running the `IF NOT EXISTS` query directly on the Hive / SQL DB vs. running it through Spark.

I think it's better to keep Spark consistent across sources, which does leave a concession for us being inconsistent in the above way. We should document the behavior of the SQL queries where they may deviate from the behavior of the underlying source where appropriate.

Reply:
Thanks, @mccheah! I talked with Wenchen this morning and I think we are all in agreement now that we should guarantee consistency.